### ISSUE TITLE
Interface to display results 

### ISSUE THREAD

**Issue by fingenerf** [pymontecarlo/pymontecarlo#35]
I'm working at a interface to display the results of a simulation at the moment.
For that I need some data to display. So there are mainly to questions for now:
1. Are there any sample results to use?
2. How should I read the data? I reckon it won't be necessary to read the HDF5 file directly. So will there be any function to do this?

**Comment by 0b11001111** [comment]
@fingenerf wants to display result data and tries to import them directly from a hdf5 file in the moment. This is a little overkill since there's probably any class to represent those data already (maybe similar to trajectories I'm going to display!?).
So what we need is some example data and a short introduction into how to import them into our code.

**Comment by drix00** [comment]
I just post an example of results data from mcxray in an hdf5 file in issue #16. You can use it as starting point if you need data for testing. The underling hdf5 format will not be the same for pymontecarlo project. but it should be easy to change later and work on the interface now.
Drobbox link: https://www.dropbox.com/s/jtw13x396ix24ym/SimulationHDF5.hdf5?dl=0

**Comment by ppinard** [comment]
> @fingenerf wants to display result data and tries to import them directly from a hdf5 file in the moment. This is a little overkill since there's probably any class to represent those data already (maybe similar to 

Yes the display of results should not be related to the file format. I am finalizing some of the result classes which should be used by the display. See the package [results](https://github.com/pymontecarlo/pymontecarlo/tree/emas2017/pymontecarlo/results).

My concept for the results is the following:
- A summary table where all results are shown, where each row is a simulation
- A 2D plot, where the user selects one quantity for the x-axis (e.g. beam energy) to plot against one or more quantities for the y-axis (e.g. X-ray intensity).

Both rely on extracting the results as a `key: value` pair. For this I created what I called a [`DataRow`](https://github.com/pymontecarlo/pymontecarlo/blob/emas2017/pymontecarlo/util/datarow.py). Compatible result classes should implement the interface [`DataRowCreator`](https://github.com/pymontecarlo/pymontecarlo/blob/emas2017/pymontecarlo/util/datarow.py#L16). The [`Options`](https://github.com/pymontecarlo/pymontecarlo/blob/emas2017/pymontecarlo/options/options.py#L16) class already implements this interface. Look at the unit test [`testcreate_datarow`](https://github.com/pymontecarlo/pymontecarlo/blob/emas2017/pymontecarlo/options/test_options.py#L29) to use what calling the method `create_datarow` does. As a Python `set`, you can *union* or *intersect* two  datarows to merge them or keep only differing `key: value` pairs.

As the 2D plot is concerned, here is what I had in mind:
![result_summary](https://cloud.githubusercontent.com/assets/715280/23682776/5facb7ae-038d-11e7-8e2a-17f414100ce2.png)

**Comment by fingenerf** [comment]
First I have a few basic questions.
1. Where should I push my first drafts to? Should I create a new branch?
2. Where should I save the file in the project?

Apart from that:

> A 2D plot, where the user selects one quantity for the x-axis (e.g. beam energy) to plot against one or more quantities for the y-axis (e.g. X-ray intensity).

What do you mean by against on or more quantities?
Do you mean that there can be several graphs, for each sample one (e.g. one for Al Ka and one for Mg Ka like in your sketch) or should the actually several graphs for each sample?

**Comment by ppinard** [comment]
> Where should I push my first drafts to? Should I create a new branch?

In [pymontecarlo-gui](https://github.com/pymontecarlo/pymontecarlo-gui), create a new branch (maybe something like `result_summary`) from my `emas2017` branch.

> Where should I save the file in the project?

I don't think the figure itself should be saved in the project. Maybe we should think of a way to store the parameters to create this figure in the project, but that is not the number one priority. To save the figure as a PNG we can use the matplotlib save function.

> Do you mean that there can be several graphs, for each sample one (e.g. one for Al Ka and one for Mg Ka like in your sketch) or should the actually several graphs for each sample?

Only one graph. There could be more than one element per sample, so we can plot both X-ray lines on the same graph as shown in my little scketch.

**Comment by ppinard** [comment]
As requested, I created a widget to present the results as a table. You can find it [here](https://github.com/pymontecarlo/pymontecarlo-gui/blob/emas2017/pymontecarlo_gui/results/summary.py). It looks like this

![screenshot from 2017-03-19 11 14 57](https://cloud.githubusercontent.com/assets/715280/24080363/9840a09c-0c95-11e7-92f6-2e66b3efbb07.png)

**Comment by fingenerf** [comment]
I had a look at the new widget and I didn't understand completely so far, but for that I will have some questions later.

But for now, what will be my task now? Creating a widget to plot this data? And how are these two widgets are combined for the user? Shouldn't there be some kind of a 'plot-button' on the summary widget?

**Comment by ppinard** [comment]
> But for now, what will be my task now?

Yes, create a widget to plot this data.

> And how are these two widgets are combined for the user?

My plan was to have a tree with the open project, showing each simulation and special items including these two widgets. We could also have a plot button...

**Comment by ppinard** [comment]
As discussed on Friday, I split the `create_datarows` method of the `Project` class in two: `create_options_datarows` and `create_results_datarows`. I also added a property `result_classes` to list all available types of result. 

To read a project, you can do as follows:

```python
from pymontecarlo.project import Project

project = Project.read(filepath)
```

I think you should have all the tools you need to create the display.

**Comment by fingenerf** [comment]
Looks good so far. I reckon I will have a first draft at some stage tomorrow.

The only thing I'm missing is something to get the different options classes to list them in a combo box, so the user can choose one for the X-Axis and after that as well something to get the datarow for the chosen options class
I reckon a function like `option_classes` to get the different option classes and maybe a function to create a datarow for a known optionclass. Or do you have any other ideas?

**Comment by ppinard** [comment]
There is only one `Option` class. The idea is

- x-axis: call `project.create_options_datarows(only_different_columns=True)`, the union of all datarows will give you the column names for the combobox. See [here](https://github.com/pymontecarlo/pymontecarlo-gui/blob/emas2017/pymontecarlo_gui/results/summary.py#L55) for an example.

- y-axis: `create_results_datarows([result_class_selected_from_the_combobox])`

**Comment by fingenerf** [comment]
Ok I reckon it works out so far.

I just made a new branch [`result_summary`](https://github.com/pymontecarlo/pymontecarlo-gui/tree/result_summary) and pushed my first draft [` 8dcff0c`](https://github.com/pymontecarlo/pymontecarlo-gui/commit/8dcff0c1bf46cb5ce17c74f65392c2b6e7936561).

I didn't add matplotlib so far but I already tried to make an model `ResultPlotModel` for it, but it isn't finished yet.
But I reckon it would be good if you can have a look at the other classes and just say if it is alright like this or if there is something fundamental wrong.

**Comment by ppinard** [comment]
The structure looks good.

I am not sure I understand why you would need a "Model". For the QTableView this is required, but for a figure I think we might want to create a class similar to [SampleFigure](https://github.com/pymontecarlo/pymontecarlo/blob/preview/pymontecarlo/figures/sample.py#L30) of issue #34. 

In the `XaxisComboBoxWidget`, I also do not understand the `setOptions` and `currentOption` methods. The combo box should take the column names from all `Options`' data rows. The same goes for the `XrayLineListWidget` which should be able to take other types than `XrayLine`.

**Comment by fingenerf** [comment]
I integrated `setOptions` into `setProject` in `XaxisComboBoxWidget` and `setXrayLines` into `setResultClass` in `XrayLineListWidget`.
But I need the `currentOption` and `xrayLines` functions to get that for the plot.
Apart from that I reckon it's the easier to talk about the plotting later directly.
For the changes see [`1f1f8bb`](https://github.com/pymontecarlo/pymontecarlo-gui/commit/1f1f8bb854af07decfbd349381fb5a1eb28a92e2)

**Comment by fingenerf** [comment]
I just pushed some new code, for that see [`fb38238`](https://github.com/pymontecarlo/pymontecarlo-gui/commit/fb382388497c24610782d17ac97cbbb0d8ac2d20).

It's working now so far.
But on the one hand the keys for the different graphs in the plot are missing and other stuff for the ploting. I will add this and some documentation soon.

On the other hand I struggled a fair bit with the error in the datarows and the units in the name of the dicts in the datarows, when I wanted to call them for the data to plot.
Apart from that I had issues with the results, when the value is 'nan', since the name is different in the different dicts in that case, e.g. `<class 'list'>: [{'Cu K–L3': nan}, {'Cu K–L3': nan}, {'Cu K–L3 (1/mol/rad²)': 10.0, 'σ(Cu K–L3) (1/mol/rad²)': 0.1}]`.
Maybe you can have a look at my solution if they make sense.

**Comment by ppinard** [comment]
Implemented in version [1.0](https://github.com/pymontecarlo/pymontecarlo/releases/tag/1.0.0)
